realm/src/main/java/io/realm/RealmQuery.java

```diff
@@ -1297,9 +1698,333 @@ public long count() {
     public E findFirst() {
         long rowIndex = this.query.find();
         if (rowIndex >= 0) {
-            return realm.get(clazz, (view != null) ? view.getTargetRowIndex(rowIndex) : rowIndex);
+            return realm.getByIndex(clazz, (view != null) ? view.getTargetRowIndex(rowIndex) : rowIndex);
         } else {
             return null;
         }
     }
+
+    /**
+     * Similar to {@link #findFirst()} but runs asynchronously from a worker thread
+     * @param callback to receive the result of this query
+     * @return A {@link Request} representing a cancellable, pending asynchronous query
+     */
+    public Request findFirst(RealmObject.QueryCallback<E> callback) {
+        // will use the Looper of the caller thread to post the result
+        final Handler handler = new EventHandler(callback);
+
+        // We need a pointer to the caller Realm, to be able to handover the result to it
+        final long callerSharedGroupNativePtr = realm.getSharedGroupPointer();
+
+        // Handover the query (to be used by a worker thread)
+        final long ptrQuery = query.handoverQuery(callerSharedGroupNativePtr);
+
+        // We need to use the same configuration to open a background SharedGroup (i.e Realm)
+        // to perform the query
+        final RealmConfiguration realmConfiguration = realm.getConfiguration();
+
+        // This call needs to be done on the caller's thread, since SG()->get_version_of_current_transaction is not thread safe
+        final long[] callerSharedGroupVersion = realm.getSharedGroupVersion();
+
+        Future<?> pendingQuery = asyncQueryExecutor.submit(new Runnable() {
+            @Override
+            public void run() {
+                if (!Thread.currentThread().isInterrupted() && (asyncRequest == null || !asyncRequest.isCancelled())) {
+                    Realm bgRealm = null;
+
+                    try {
+                        //TODO Once SharedGroup is thread safe, start reusing a cached instance
+                        //     of a worker Realm to avoid the cost of opening/closing a SharedGroup
+                        //     for each query
+                        bgRealm = Realm.getInstance(realmConfiguration);
+
+                        // begin_read may throw an UnreachableVersionException if the provided version
+                        // is no longer available, we fail fast
+                        bgRealm.setSharedGroupAtVersion(callerSharedGroupVersion);
+
+                        // Run the query & handover the table view for the caller thread
+                        long rowPtr = query.findWithHandover(bgRealm.getSharedGroupPointer(), ptrQuery);
+
```

Comment:
Would the rowPtr be leaked if the task got cancelled here since it is not wrapped by the Reference yet?